Content/Items/Breacher/Projectiles.BreachCannonSentry.cs

```diff
@@ -0,0 +1,585 @@
+
+			spriteBatch.Begin(default, default, default, default, default, default, Main.GameViewMatrix.TransformationMatrix);
+		}
+
+		public void DrawPrimitives()
+		{
+			Effect effect = Filters.Scene["BreachLaser"].GetShader().Shader;
+
+			Matrix world = Matrix.CreateTranslation(-Main.screenPosition.Vec3());
+			Matrix view = Main.GameViewMatrix.ZoomMatrix;
+			Matrix projection = Matrix.CreateOrthographicOffCenter(0, Main.screenWidth, Main.screenHeight, 0, -1, 1);
+
+			effect.Parameters["transformMatrix"].SetValue(world * view * projection);
+			effect.Parameters["sampleTexture"].SetValue(ModContent.Request<Texture2D>("StarlightRiver/Assets/GlowTrail").Value);
+			effect.Parameters["noiseTexture"].SetValue(ModContent.Request<Texture2D>("StarlightRiver/Assets/Noise/ShaderNoiseLooping").Value);
+
+			effect.Parameters["time"].SetValue(Main.GameUpdateCount * 0.25f);
+			effect.Parameters["stretch"].SetValue(2f / laserLength);
+
+			effect.Parameters["dilation"].SetValue(0.8f);
+			effect.Parameters["falloff"].SetValue(1);
+
+			trail2?.Render(effect);
+			if (superLaser)
+				superTrail2?.Render(effect);
+		}
+
+		private bool InvertRotation()
+		{
+			Vector2 originVector = Projectile.rotation.ToRotationVector2();
+
+			switch (originAngle)
+			{
+				case 0:
+					return originVector.Y > 0;
+				case 1:
+					return originVector.X < 0;
+				case 2:
+					return originVector.Y < 0;
+				case 3:
+					return originVector.X > 0;
+			}
+
+			return false;
+		}
+
+		private Vector2 CalculateOffset(Vector2 start, float rot, int width, int pierce)
+        {
+			if (start == Vector2.Zero)
+				return Vector2.One;
+
+			Vector2 offset = Vector2.Zero;
+
+			for (int k = 0; k < 50; k++)
+			{
+				offset = rot.ToRotationVector2() * k * 16;
+
+				int i = (int)((start.X + offset.X) / 16);
+				int j = (int)((start.Y + offset.Y) / 16);
+				Tile testTile = Main.tile[i, j];
+
+				if (testTile.HasTile && Main.tileSolid[testTile.TileType] && !TileID.Sets.Platforms[testTile.TileType])
+				{
+					break;
+				}
+			}
+
+			Vector2 testEndpoint = start + offset;
+			NPC[] sortedNPC = Main.npc.Where(n => n.active && !n.friendly && !n.CountsAsACritter).OrderBy(n => (n.Center - start).Length()).ToArray();
+
+			if (sortedNPC.Length == 0)
+				return testEndpoint - start;
+
+			int pierceLeft = pierce;
+
+			for (int index = 0; index < sortedNPC.Length; index++)
+			{
+				NPC npc = sortedNPC[index];
+				float collisionPoint = 0f;
+
+				if (Collision.CheckAABBvLineCollision(npc.Hitbox.TopLeft(), npc.Hitbox.Size(), start, testEndpoint, width, ref collisionPoint))
+				{
+					pierce--;
+					if (pierce <= 0)
+                    {
+						testEndpoint = Vector2.Lerp(start, testEndpoint, (float)(collisionPoint + 8) / (start - testEndpoint).Length());
+						break;
+					}
+				}
+			}
+
+			return testEndpoint - start;
+		}
+
+		private void DrawBalls(SpriteBatch sb, BlendState endState, Color topColor, Color bottomColor, float scale)
```

Comment:
Shouldn't this be ``DrawBall`` as it only appears to draw a single ball?

Reply:
It draws both the main laser ball and the superlaser ball (if applicable)